marketplace/google-analytics-app/src/AppConfig.tsx

```diff
@@ -0,0 +1,342 @@
+    const { contentTypes, clientId, viewId } = this.state;
+    const { notifier } = this.props.sdk;
+
+    if (!clientId || !viewId) {
+      notifier.error('You must provide both a valid client ID and view ID!');
+
+      return false;
+    }
+
+    if (!/^[-a-z0-9]+.apps.googleusercontent.com$/i.test(clientId)) {
+      notifier.error("The value given for the client ID doesn't look valid!");
+      return false;
+    }
+
+    const ctKeys = Object.keys(contentTypes);
+
+    if (!ctKeys.length || !ctKeys[0]) {
+      notifier.error('You need to select at least one content type with a slug field!');
+    }
+
+    if (ctKeys.some(key => key && !contentTypes[key].slugField)) {
+      notifier.error('Please complete or remove the incomplete content type rows!');
+      return false;
+    }
+
+    const editorInterface = ctKeys.reduce(
+      (acc: { [key: string]: Partial<EditorInterface> }, id) => {
+        const sidebarPosition: { [key: string]: object } = { sidebar: { position: 1 } };
+
+        acc[id] = sidebarPosition;
+        return acc;
+      },
+      {}
+    );
+
+    return {
+      parameters: {
+        contentTypes,
+        clientId,
+        viewId
+      },
+      targetState: {
+        EditorInterface: editorInterface
+      }
+    };
+  }
+
+  handleContentTypeChange(prevKey: string, newKey: string) {
+    this.setState(prevState => {
+      const contentTypes: AppConfigState['contentTypes'] = {};
+
+      // remove contentType[prevKey] field and replace with the new contentType
+      // key while preserving key order
+      for (const [prop, value] of Object.entries(prevState.contentTypes)) {
+        if (prop === prevKey) {
+          contentTypes[newKey as keyof typeof contentTypes] = {
+            slugField: '',
+            urlPrefix: value.urlPrefix
+          };
+        } else {
+          contentTypes[prop] = value;
+        }
+      }
+
+      return {
+        contentTypes
+      };
+    });
+  }
+
+  handleContentTypeFieldChange(key: string, field: string, value: string) {
+    this.setState(prevState => {
+      const prevContentTypes = prevState.contentTypes;
+      const curContentTypeProps = prevContentTypes[key];
+
+      return {
+        contentTypes: {
+          ...prevState.contentTypes,
+          [key]: {
+            ...curContentTypeProps,
+            [field]: value
+          }
+        }
+      };
+    });
+  }
+
+  addContentType() {
+    this.setState(prevState => ({
+      contentTypes: {
+        ...prevState.contentTypes,
+        '': { slugField: '', urlPrefix: '' }
+      }
+    }));
+  }
+
+  removeContentType(key: string) {
+    this.setState(prevState => {
+      const contentTypes = { ...prevState.contentTypes };
+
+      delete contentTypes[key];
+
+      return { contentTypes };
+    });
+  }
+
+  render() {
+    const { contentTypes, allContentTypes } = this.state;
+
+    const optionPlaceholderAttrs: { [key: string]: boolean } = { disabled: true, selected: true };
```

Comment:
This looks like something that could live outside of the component. And the typing I think is not needed, should be inferred.

Reply:
The problem here was that the `<Options>` component from forma doesn't have props for those attributes on the HTMLOptionElement. It turns out regular `<option>` elements work as children to forma's `<Select>` component so I've now used those.